# Coding Session 01
## Getting started with campaign data

Today we will use Python to work out what a campaign budget can buy, change our assumptions, and explain the results. No previous coding experience is needed.

## Before we code: AI and the work of an analyst

What does AI mean for the skills we should develop in this course? The following questions and answers are reproduced verbatim from Shiro Kuriwaki’s slides reporting August 2026 interviews with a **Project Lead at Boston Consulting Group** and a **Government Official**. Line breaks and bullet formatting have been adjusted for this notebook.

**Source:** Shiro Kuriwaki, *Applied Quantitative Research Design*, Class 0.1 (September 2, 2026), slides 19–22, “Employers on AI (Aug. 2026 interview).” These are the excerpts presented in the slides, not a full interview transcript.

---

### 01 · Most common mistakes for new graduates?

| <img src="https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/interview_consulting.svg?raw=1" width="64" alt="Consulting briefcase icon"><br>Project Lead<br>Boston Consulting Group | <img src="https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/interview_government.svg?raw=1" width="64" alt="Government building icon"><br>Government Official |
| :--- | :--- |
| “Not enough sensitivity checks on the data” they present<br><br>“Even descriptive stats of the data seems basic but is nontrivial in the consulting world” … | “Senior policymakers operate under immense time pressure.”<br><br>“Data helps but too much becomes overwhelming.” |


---

### 02 · How has AI changed your work?

| <img src="https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/interview_consulting.svg?raw=1" width="64" alt="Consulting briefcase icon"><br>Project Lead<br>Boston Consulting Group | <img src="https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/interview_government.svg?raw=1" width="64" alt="Government building icon"><br>Government Official |
| :--- | :--- |
| Powerful AI tools → “Standards in consulting have honestly become higher.”<br><br>Managers now expect entry-level consultants to “use AI to get data analysis done quickly, correctly, and efficiently” | “certain [repetitive] functions have really sped up”<br><br>But “a lot of our work requires judgement — knowledge of what has happened on the sidelines of in-person meetings, who knows whom — which AI cannot know. ” |


---

### 03 · What distinguishes an excellent vs. average employee?

| <img src="https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/interview_consulting.svg?raw=1" width="64" alt="Consulting briefcase icon"><br>Project Lead<br>Boston Consulting Group | <img src="https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/interview_government.svg?raw=1" width="64" alt="Government building icon"><br>Government Official |
| :--- | :--- |
| People who excel in their first year are “really creative in using AI to sanity check their output, test different ways for data visualizations, to tell the story … but also in a way not to manipulate the data so much” | “Junior economists struggle to distinguish insight from information. It takes time to develop intuition to know what kind of data is convincing.” |


## Why these coding sessions matter in the age of AI

AI can produce code, tables, charts, and explanations. To use them well, we need to understand the code, read the data, interpret the visualizations, and decide what deserves to be presented. **These coding sessions are designed to build those skills.**

| What we need to do | What we will practice |
| :--- | :--- |
| **Work quickly, on the fly** | Read code, change an input, rerun an analysis, and check the result |
| **Decide what to present and how** | Select the comparison that matters and choose a clear number, table, or chart. Check axes, scales, units, and uncertainty before presenting a visualization. |
| **Turn information into insight** | Explain what a result tells us about the campaign question and what action it supports. Keep the evidence that matters and remove what distracts. |
| **Do better than AI alone** | Challenge its answers, catch misleading comparisons, test assumptions, and improve the explanation. Be able to defend why the final analysis is more trustworthy and useful than the first generated output. |


## Today's plan

| What we will practice |
| :--- |
| Lab goals and using AI thoughtfully |
| Saving a notebook, running cells, and reading output |
| Arithmetic, variables, and campaign costs |
| Changing assumptions and fixing errors |
| Reading a small table and calculating a column |
| Independent practice and interpretation |
| Checking our work and questions |

By the end, you should be able to calculate affordable campaign activity and average cost, rerun a calculation after changing an input, and explain your answer with the correct units.

### Where notebooks, editors, and the CLI fit

The same Python calculation can run through different interfaces. We will use **Colab notebooks** in these labs. An **IDE** (integrated development environment) combines editing, running, and debugging tools. A **CLI** (command-line interface) lets you give commands by typing in a terminal.

![Coding interfaces and AI agents: Codex and Claude Code can read files, edit code, and run checks for human review. Colab runs notebook cells; an editor or IDE manages and runs code files; a terminal provides a command-line interface to run a saved script. All use a Python runtime, which may be local or remote.](https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/coding_environments.png?raw=1)

An interface is where you interact with code; the **runtime** is what executes it. An editor may contain both notebooks and a terminal.

### Coding agents: Codex and Claude Code

**Codex (OpenAI)** and **Claude Code (Anthropic)** are examples of AI coding agents. They can inspect project files, make code changes, and run commands or checks, within the access you give them. You can work with them through terminal CLIs or editor integrations. See the official [Codex CLI](https://learn.chatgpt.com/docs/codex/cli), [Codex editor integration](https://learn.chatgpt.com/docs/codex/ide), and [Claude Code overview](https://code.claude.com/docs/en/overview).

We will use Colab for this session. You do not need to install either agent; any use for assignments must follow the course AI rules.

## Accessing AI tools at Yale

*Access information checked September 8, 2026.*

### Available now: Yale Clarity

Open [Clarity](https://clarity.yale.edu/) and sign in with your Yale account. Yale provides free access to the chat models available in your account, including GPT and Claude models. You can use them for explanations and coding questions where course rules permit. See [Yale’s Clarity overview](https://ai.yale.edu/yales-ai-tools-and-resources/clarity-platform).

**Clarity and Codex are different tools.** Access to GPT models in Clarity does not establish access to the Codex coding agent. Yale’s [current tool guidance](https://poorvucenter.yale.edu/teaching/teaching-resource-library/ai-guidance-for-teachers/ai-tools/yale-provided-tools) says dedicated coding agents such as Codex and Claude Code are not currently supported.


### Expected in mid-September: a Claude offering

Shiro’s slide announces a Claude offering expected to launch in **mid-September 2026**. Treat this as an announced expectation; check Yale’s launch instructions for the confirmed date, eligibility, cost, and whether Claude Code is included. Claude chat models are already available through Clarity.

![](https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/shiro_yale_ai_access_slide16.png?raw=1)


We will use Colab for this session; no additional AI subscription is required.

## 1. Your notebook

### Student Colab guide

The guide below covers opening an assignment, saving your own copy, running cells, writing answers, submitting work, and troubleshooting.

[Open or download the complete student Colab guide](assets/student_colab_guide.pdf).


![Colab guide page 1: vocabulary, opening the assignment, and saving your own copy in Drive.](https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/student_colab_guide/page-01.png?raw=1)

![Colab guide page 2: Save a copy in Drive menu screenshot and running the setup cell.](https://github.com/albertostefanelli/DSPC_coding_sessions/blob/master/codingsession01/assets/student_colab_guide/page-02.png?raw=1)


### Try the notebook

Google Colab runs Python in your browser. A notebook combines **code cells**, which execute Python, and **text cells**, which contain instructions and explanations.

1. Save your own copy in Drive and work in that copy.
2. Edit the text cell below by double-clicking it and adding your name.
3. Run a code cell with the play button or `Shift + Enter`.
4. Add a new code cell and write a short greeting using `print(...)`.

Save your work as you go.

**Your name:** Replace this text with your name.

In [ ]:
print("Welcome to campaign data science!")

### What appears below a cell?

Python executes lines from top to bottom. A notebook automatically displays the value of the **last expression** in a cell. Earlier expressions still run, but their values are not automatically shown.

**Predict before running:** Which result will appear?

In [ ]:
2 + 3
10 - 4

Use `print(...)` when you want to display more than one result. Each call below displays its output when that line runs.

In [ ]:
print(2 + 3)
print(10 - 4)

## 2. Python as a calculator

| Operator | Meaning | Example |
|---|---|---|
| `+` | Add | `8 + 2` |
| `-` | Subtract | `8 - 2` |
| `*` | Multiply | `8 * 2` |
| `/` | Divide | `8 / 2` |
| `**` | Raise to a power | `8 ** 2` |

Parentheses control which calculation happens first.

Suppose a campaign has **USD 24,000**, an ad costs ** 6,000** to produce, and delivery costs **0.10 per impression**. How many impressions can the campaign afford after paying for production?

In [ ]:
# Dollars remaining after production, divided by dollars per impression.
(24000 - 6000) / 0.10

**Quick check:** Explain why `24000 - 6000 / 0.10` would answer a different question. Identify which operation Python would perform first.

## 3. Give your inputs names

An assignment evaluates the expression on the right of `=` and gives its value the name on the left. We can use that name in later calculations. Assignment alone does not display the value; use the name as the last expression or call `print`.

Use descriptive names with letters, digits, and underscores. Names cannot start with a digit or contain spaces, and capitalization matters: `budget` and `Budget` are different names. A `#` starts a comment for human readers.

In [ ]:
budget = 24000
fixed_cost = 6000
marginal_cost_per_impression = 0.10
budget

24000

### Fixed, marginal, and average cost

- **Fixed cost:** The setup expense, paid before any impressions are delivered.
- **Marginal cost:** The additional cost of delivering one more impression.
- **Average cost:** Total spending divided by impressions delivered, including setup.

We assume the full budget is spent, the budget exceeds the fixed cost, and marginal cost stays positive and constant. These practice values give whole numbers of impressions.

In [ ]:
spend_available_for_delivery = budget - fixed_cost
impressions = spend_available_for_delivery / marginal_cost_per_impression
average_cost_per_impression = budget / impressions

print("Impressions delivered:", impressions)
print("Average dollars per impression:", average_cost_per_impression)

Impressions delivered: 180000.0
Average dollars per impression: 0.13333333333333333


### Calling built-in functions

A function call uses a name followed by parentheses and arguments. We have used `print` to display results. `round(number, decimal_places)` returns a rounded number; it is useful for readable output.

Keep the unrounded value for calculations and round when displaying it.

In [ ]:
print("Marginal dollars per impression:", marginal_cost_per_impression)
print("Average dollars per impression:", round(average_cost_per_impression, 4))

Marginal dollars per impression: 0.1
Average dollars per impression: 0.1333


**Discuss:** Why is average cost about 0.1333 when the marginal cost is 0.10? Where does the difference come from?

## 4. Change an assumption

**Predict:** If the budget rises to USD 36,000 but both costs stay the same, will the number of impressions rise or fall? What will happen to average cost?

Run the next cell. Notice that assigning a new budget does **not** recalculate the previously assigned value of `impressions`. Variables hold calculated values, not live spreadsheet formulas.

In [ ]:
budget = 36000
print("New budget:", budget)
print("Previously calculated impressions:", impressions)

Now rerun the dependent calculations. You can rerun their earlier cell or run the version below. The result should reflect the new budget.

In [ ]:
spend_available_for_delivery = budget - fixed_cost
impressions = spend_available_for_delivery / marginal_cost_per_impression
average_cost_per_impression = budget / impressions

print("Updated impressions:", impressions)
print("Updated average dollars per impression:", round(average_cost_per_impression, 4))

**Check:** You should get 300,000 impressions at an average cost of 0.12 each. The fixed cost is spread across more impressions. Marginal cost has not changed.

### **Your runtime remembers what you ran**

The runtime is the program executing your notebook. It remembers definitions in the order you execute cells, which may differ from the order they appear on the page. Old displayed output can remain even when it no longer matches your inputs.

After a restart, definitions are cleared: rerun the notebook from the beginning. If a computation gets stuck, interrupt it; if necessary, restart the session and rerun your cells.

### Debugging practice

Errors are a normal part of coding. Read the final line of the error message, locate the relevant code, and make one change at a time.

The broken lines below are commented out so the notebook can run from start to finish. For each one: predict the problem, remove the leading `#`, run it, then fix the line. Leave the corrected version in place.

In [ ]:
# Practice A: remove the leading #, run, then repair the missing punctuation.
# print("Campaign budget:", budget

In [ ]:
# Practice B: remove the leading #, run, then check the variable's capitalization.
# print("Campaign budget:", Budget)

## 5. Read the same kind of information from a table

The Week 1 lab has a small CSV file of campaign tactics. A **DataFrame** is a table: each row is a tactic, and each column describes an attribute of that tactic.

`import pandas as pd` loads the table tools and gives them the short name `pd`. `pd.read_csv(...)` reads the course file directly; you do not need to download or upload it. This cell needs an internet connection.

In [ ]:
import pandas as pd

course_data_url = (
    "https://raw.githubusercontent.com/albertostefanelli/DSPC_coding_sessions/"
    "master/codingsession01/data/tactics.csv"
)
tactics = pd.read_csv(course_data_url)
tactics

,tactic,fixed_cost,marginal_cost_per_unit,unit_label,units_at_50k
0,mailer,0,0.50,household,100000
1,digital_ad,30000,0.05,impression,400000


### Select columns from a table

Often we only need a few columns. Use square brackets and put the column names in quotation marks. Names must match the table exactly, including capitalization and underscores.

**One column:** `tactics["tactic"]` selects the tactic names. The result is a **Series**, a single labeled column.

In [ ]:
tactics["tactic"]

**Several columns:** Put a list of names inside the brackets. The inner brackets make the list; the outer brackets select those columns from `tactics`.

The result is a **DataFrame**, a table. Columns appear in the order you request. This selection does not change the original `tactics` table.

In [ ]:
tactics[["tactic", "fixed_cost", "marginal_cost_per_unit"]]

**Try it:** In the cell below, select only `tactic` and `unit_label`, in that order. Use two pairs of square brackets and separate the quoted names with a comma.

Notice the difference: `tactics["tactic"]` returns a Series, while `tactics[["tactic"]]` keeps a one-column DataFrame.

In [ ]:
# Select the tactic and unit_label columns here.

**Look first:** Find the tactic names, fixed costs, marginal costs, and units. Which row measures households? Which measures impressions?

`tactics["fixed_cost"]` selects a column. Arithmetic on columns applies the calculation to each row. The next cell creates `average_cost` using the same idea as before:

**Average cost = (fixed cost + marginal cost × delivered units) / delivered units.**

Here we use the table's `units_at_50k` column.

In [ ]:
tactics["average_cost"] = (
    tactics["fixed_cost"]
    + tactics["marginal_cost_per_unit"] * tactics["units_at_50k"]
) / tactics["units_at_50k"]

tactics[["tactic", "unit_label", "marginal_cost_per_unit", "average_cost"]]

**Interpret:** Why can't we select the best tactic simply by choosing the smallest average cost in this table?

Keep the units in your explanation. Impressions and households are different measures of campaign activity; neither directly tells us how many votes a tactic gains.

## 6. Your turn

A campaign is considering a new canvassing program. Use these fictional practice inputs:

- Budget: **30,000**.
- Fixed training cost: **6,000**.
- Marginal cost: **12 per completed conversation**.

1. Create descriptive variables for the three inputs.
2. Calculate affordable conversations and average cost per conversation.
3. Print both results with labels.
4. In the second code cell, reduce the fixed cost to **3,000** and recalculate. Predict the direction of the changes before running it.
5. In the text cell, explain the change in two or three sentences. Use **fixed cost**, **marginal cost**, and **average cost**. Name one more piece of information you would need to recommend canvassing over digital ads.

Work through **understand → plan → execute → check**. Start with your own attempt, then ask for a hint if you get stuck.

In [ ]:
# Your code: define the inputs, calculate conversations and average cost,
# and print your results with units.

In [ ]:
# Your code: change the fixed training cost to 3000, then recalculate.
# Write your prediction as a comment before running this cell.

**Your explanation:** Replace this text with your answer.

<details>
<summary>Hints if you need them</summary>

1. Subtract the fixed cost from the budget to find the money available for conversations.
2. Divide that amount by the marginal cost per conversation.
3. Divide total spending by conversations to calculate average cost.
4. Check that fixed cost plus delivery spending equals your budget.

</details>

## Questions?

Bring unfinished questions to office hours. You can also email Alberto at alberto.stefanelli@yale.edu.